packages/assets-controllers/src/RatesController/RatesController.ts

```diff
@@ -0,0 +1,186 @@
+import { Mutex } from 'async-mutex';
+
+import { fetchMultiExchangeRate as defaultFetchExchangeRate } from '../crypto-compare-service';
+import type {
+  RatesMessenger,
+  ConversionRates,
+  RatesControllerState,
+  RatesControllerOptions,
+} from './types';
+
+export const name = 'RatesController';
+
+export enum Cryptocurrency {
+  Btc = 'btc',
+}
+
+const DEFAULT_INTERVAL = 180000;
+
+const metadata = {
+  currency: { persist: true, anonymous: true },
+  rates: { persist: true, anonymous: true },
+  fromCurrencies: { persist: true, anonymous: true },
+};
+
+const defaultState = {
+  currency: 'usd',
+  rates: {
+    [Cryptocurrency.Btc]: {
+      conversionDate: 0,
+      conversionRate: '0',
+    },
+  },
+  fromCurrencies: [Cryptocurrency.Btc],
+};
+
+export class RatesController extends BaseController<
+  typeof name,
+  RatesControllerState,
+  RatesMessenger
+> {
+  readonly #mutex = new Mutex();
+
+  readonly #fetchMultiExchangeRate;
+
+  readonly #includeUsdRate;
+
+  #intervalLength: number;
+
+  #intervalId: NodeJS.Timeout | undefined;
+
+  /**
+   * Creates a RatesController instance.
+   *
+   * @param options - Constructor options.
+   * @param options.includeUsdRate - Keep track of the USD rate in addition to the current currency rate.
+   * @param options.interval - The polling interval, in milliseconds.
+   * @param options.messenger - A reference to the messaging system.
+   * @param options.state - Initial state to set on this controller.
+   * @param options.fetchMultiExchangeRate - Fetches the exchange rate from an external API. This option is primarily meant for use in unit tests.
+   */
+  constructor({
+    interval = DEFAULT_INTERVAL,
+    messenger,
+    state,
+    includeUsdRate,
+    fetchMultiExchangeRate = defaultFetchExchangeRate,
+  }: RatesControllerOptions) {
+    super({
+      name,
+      metadata,
+      messenger,
+      state: { ...defaultState, ...state },
+    });
+    this.#includeUsdRate = includeUsdRate;
+    this.#fetchMultiExchangeRate = fetchMultiExchangeRate;
+    this.#intervalLength = interval;
+  }
+
+  async #withLock<R>(f: () => R) {
+    const releaseLock = await this.#mutex.acquire();
+    try {
+      return f();
+    } finally {
+      releaseLock();
+    }
+  }
+
+  /**
+   * Executes the polling operation to update rates.
+   */
+  async #executePoll(): Promise<void> {
+    await this.updateRates();
+  }
+
+  /**
+   * Updates the rates by fetching new data.
+   */
+  async updateRates(): Promise<void> {
+    await this.#withLock(async () => {
+      const { currency, fromCurrencies } = this.state;
+      const response = await this.#fetchMultiExchangeRate(
+        currency,
+        fromCurrencies,
+        this.#includeUsdRate,
+      );
+
+      const updatedRates: ConversionRates = {};
+      for (const [cryptocurrency, values] of Object.entries(response)) {
+        updatedRates[cryptocurrency] = {
+          // Divided by 1000 to convert to ms
```

Comment:
Doesn't `Date.now()` already return the number in milliseconds? It seems like we are converting this to seconds here. If so, why?

Reply:
Good catch, this comment is wrong because of my previous review... I misread the doc and thought we were converting to `ms` here, but I've checked the doc, and `Date.now()` is already milliseconds yup! 🙈 

Maybe we should keep this as `ms` yes.

Reply:
Good catch, missed this.

I can't really say the exact reason but this is how the time in handled in the `CurrencyRatesController`. I think I'll just return ms and let the consumer to the required convertions

Reply:
Done in https://github.com/MetaMask/core/pull/4242/commits/bf52ebe716e62ce5c5839503936256fc7bccdd00